### PR TITLE
Dev.casing

### PR DESCRIPTION
**Pull request by roedoejet**
Here's an attempt to address https://github.com/roedoejet/g2p/issues/270

I don't really understand the `kwk` mapping, so I just tried to address it in a general case. Curious to discuss with @joanise to see if it works sufficiently for what's needed for `kwk`.

### CODE REVIEW

File: g2p/transducer/__init__.py
Comment (joanise) [inline g2p/transducer/__init__.py:1264]: I guess it's OK to assume this is a 1-1 mapping, as you do here.

File: g2p/tests/test_transducer.py
Comment (joanise) [inline g2p/tests/test_transducer.py:234]: What do we do in the opposite scenario? kwk has complexities that I'd like to know are addressed. E.g., I think this is what we want:

context based:

    ᴇTO -> ETO
    ᴇto -> eto

1 char -> 2 char context based:

    ʟat -> tłat
    ʟAT -> TŁAT

2 char -> 3 char context based:

    EᴇT -> EʼA̱T
    Eᴇt -> E'a̱t

and I'll want to test what preserve_case does in all these cases when we turn it on for kwk BOAS->Umista.

File: g2p/transducer/__init__.py
Comment (joanise) [inline g2p/transducer/__init__.py:1277]: So you're assuming case_equivalencies only lists complete rule outputs, rather than characters? What about `ʟ -> tλ` which would rely on the standard "t":"T" but would have to declare "λ":"⅄"?

File: g2p/transducer/__init__.py
Comment (joanise) [inline g2p/transducer/__init__.py:1306]: Boy, it's hard to follow all this logic! We need more extensive unit testing just to make sure it works.

E.g., exercise these:
 - Given rule `aa -> bb`, `Aaxyz` should get mapped to `Bbxyz`
 - the hard stuff in kwk BOAS with those small caps being mapped to a context-dependent case
 - Given rule `a -> bb` `AXYZ` should get mapped to `BBXYZ` but `Axyz` to `Bbxyz` (although this is actually not fully reliable, e.g., in Spanish in some cases `ll` is capitalized as a unit, e.g. `llama` might capitalize to `LLama` though from my experience that's not a uniformly applied rule. And does kwk Umista consider `tł` as a single letter capitalized jointly, or as two letters with only the T being uppercased in capitalized words?)

File: g2p/tests/test_transducer.py
Comment (joanise) [inline g2p/tests/test_transducer.py:228]: You probably meant to lowercase one of these two lines
```suggestion
        self.assertEqual(transducer("e\u0301").output_string, "f")
```

File: g2p/tests/test_transducer.py
Comment (joanise) [inline g2p/tests/test_transducer.py:230]: We can help the reader of the code by inserting the rendered characters
```suggestion
        # Test what happens in Heiltsuk. \u03BB (λ) should be capitalized as \u2144 (⅄)
```

**Review by joanise — COMMENTED** [review]
I'm not sure this handles all the hard cases correctly. Leaving a bunch of comments now, then I'll want to apply preserve case to kwk BOAS->Umista and test to see what happens with my unit test cases for that, and then I'll probably have more feedback.

At the very least this code needs quite a bit more unit testing to capture more tricky cases. See inline comments.

File: g2p/transducer/__init__.py
Comment (roedoejet) [reply g2p/transducer/__init__.py:1277]: yes because we don't really have a notion of characters anywhere (re: the tokenization issue from before). here you would have to do `{'tλ': 'T⅄'}` I suppose

File: g2p/transducer/__init__.py
Comment (roedoejet) [reply g2p/transducer/__init__.py:1306]: > Boy, it's hard to follow all this logic! We need more extensive unit testing just to make sure it works.
> 
> E.g., exercise these:
> 
> * Given rule `aa -> bb`, `Aaxyz` should get mapped to `Bbxyz`

Yes, this works.

> * the hard stuff in kwk BOAS with those small caps being mapped to a context-dependent case

I don't fundamentally really understand what is going on here, and I can't reproduce the examples you have above in the rules even without changing anything so we need to talk more about this.

> * Given rule `a -> bb` `AXYZ` should get mapped to `BBXYZ` but `Axyz` to `Bbxyz` (although this is actually not fully reliable, e.g., in Spanish in some cases `ll` is capitalized as a unit, e.g. `llama` might capitalize to `LLama` though from my experience that's not a uniformly applied rule. And does kwk Umista consider `tł` as a single letter capitalized jointly, or as two letters with only the T being uppercased in capitalized words?)

I disagree that for a rule a -> bb Axyz should get turned to Bbxyz. We will handle this as BBxyz, otherwise someone would need to add a case equivalency for bb and Bb, in which case it should work.

Comment (joanise) [reply g2p/transducer/__init__.py:1306]: I guess we might need to ask Daisy what she thinks is right for kwk. And I guess we should also test with kwk and see what comes out with your current implementation.

So for kwk my assumption has been that you can infer whether a small cap L or a small cap E should be interpreted as upper case by the case of the next letter. I suppose we should validate that assumption with Daisy before investing too much development effort making it work...

**Comment by joanise** [comment]
I'm playing with case_equivalencies now, to enhance the kwk mapping.
For example, for rule `ᴇa,a̱ʼa`, I'd like to declare the equivalency `"a̱ʼa": "A̱ʼA", so that `TᴇAtᴇa` could get mapped to `TA̱ʼAta̱ʼa` rather than `Ta̱ʼAta̱ʼa` but I can't figure out how to make it happen.
A challenge, here, is that for this kwk mapping we'll have to have declare a couple dozen equivalencies to make all the vowel pairs that include `ᴇ` work correctly.

**Comment by joanise** [comment]
@roedoejet 
Rebased and improved. While still nor perfect, I think it's time to merge this as good enough, assuming CI passes. It will help most of the time, and the errors in case preservation for kwk will just mean once in a while, a character that should have been case-preserved is not. I think it's better to accept this partial solution than do the perfect solution I was hoping for.

But boy, rebasing this over the pydantic update was not trivial! Please review my work carefully.

**Comment by joanise** [comment]
Hum, I can't request a review from you @roedoejet because it was your PR in the first place. Review it anyway and comment here.

File: g2p/tests/test_mappings.py
Comment (roedoejet) [inline g2p/tests/test_mappings.py:191]: this is a good compromise

File: g2p/tests/test_transducer.py
Comment (roedoejet) [inline g2p/tests/test_transducer.py:242]: I think these tests are pretty clear with what our implementation is

File: g2p/mappings/utils.py
Comment (roedoejet) [inline g2p/mappings/utils.py:663]: I would like to see some tests about interactions between `case_sensitive` and `preserve_case`. Does `case_sensitive=False` and `preserve_case=True` work for example?

**Review by roedoejet — COMMENTED** [review]
Looks good to me - I can't approve it, since it's my own PR. My only remaining question is the one about interactions between `case_sensitive` and `preserve_case`

File: g2p/mappings/utils.py
Comment (joanise) [reply g2p/mappings/utils.py:663]: Hum, good question. It actually only makes sense to have `case_sensitive=False` when you set `preserve_case=True`. Having both be true should raise an error.

Comment (joanise) [reply g2p/mappings/utils.py:663]: The semantics of preserve case is that case does affect the meaning of the characters (i.e., the mapping is case insensitive) but should be preserved because it has non-phonetic meaning (e.g., proper name, beginning of the sentence).

Comment (joanise) [reply g2p/mappings/utils.py:663]: So I just added a model validator that raises when the two are set incompatibly.

**Review by joanise — APPROVED** [review]

**Pull request merged**
